### PR TITLE
Fix Weave Models

### PR DESCRIPTION
**Pull request by rbharath**
The `WeaveModel` has a number of inconsistencies between it's implementation and the source paper. This PR is working towards fixing these issues.

### CODE REVIEW

**Comment by rbharath** [comment]
This PR makes a number of fixes:

- The `WeaveModel` implementation in DeepChem differed a number of ways from the implementation in the source paper. Here are the modifications made to our implementation:
  - Allow for fully connected layers at the end post `WeaveGather` as in the source paper
  - Allow for batch-normalizing after all convolutional and fully connected layers as in paper
  - Our `WeaveGather` layer behaved in a non-standard way by compressing its outputs back to the size of the inputs. On the other hand, the source paper didn't perform this compression. I've added a option in which controls this compression option.
- I've added in a lot more documentation for `WeaveModel`, `WeaveLayer`, and `WeaveGather`.
- I've added in a fix for the excess logging reported in #2023 by swapping `featurize` for `_featurize` in the places noted in the issue.

I'm not sure yet if this will resolve the benchmarking issues noted in #2029 but I believe at this point, our implementation closely matches the implementation from Google's original paper. Note that it's still possible to recover the original DeepChem implementation by setting parameters accordingly (turn off batch norm, remove fully connected layers).

**Comment by rbharath** [comment]
Travis is passing (except for a `test_roulette` failure). I'm going to go ahead and merge this in to facilitate additional benchmarking of `WeaveModel`. I might have to do another round or two of iteration on `WeaveModel` in case I missed anything
